### PR TITLE
Enable noUncheckedIndexedAccess for FF repo

### PR DESCRIPTION
**Pull request by RishhiB**
Minimum file changes needed to enable noUncheckedIndexedAccess for FF repo
We are enabling `noUncheckedIndexedAccess` to protect us from changing a type of a record where there was a named property and reduce the risk of runtime errors by ensuring that indexed access on objects are properly checked for undefined values.

Arrays are impacted by this and must be annotated even though it is not the reason for using the setting

[AB#8216](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/8216)

### CODE REVIEW

File: packages/drivers/odsp-driver/src/test/tsconfig.json [hunk 1]
```diff
@@ -12,4 +13,3 @@
 			},
 		],
 	}
-
```
Comment (jason-ha) [inline packages/drivers/odsp-driver/src/test/tsconfig.json:15]: accidental end line removal

**Review by jason-ha — APPROVED** [review]
Please fix the new line change, then proceed.
It would be okay to set false in test base file if we are adjusting those.
Efficiency note: could have done both of the settings like this at once.

File: packages/drivers/odsp-driver/src/test/tsconfig.json (hunk 1 above)
Comment (RishhiB) [reply packages/drivers/odsp-driver/src/test/tsconfig.json:15]: its not letting me add it :(

Comment (jason-ha) [reply packages/drivers/odsp-driver/src/test/tsconfig.json:15]: What isn't letting you?
Actually, I think this is probably a trick of the diff UI. When I look at the file locally I see that it is improperly indented. The last line has a tab. So, I expect that locally you have tab removed but still a blank line and that causes the funny UI. I checked with another diff tool and see it is tab removal:
image
But also I see there is extra indentation. Can you remove the extra indent for the file (either now or in next PR)?

**Comment by jason-ha** [comment]
Most of our jest tests are configured to run against the .ts code rather than be built by tsc first. There are compile errors in three packages:
image

example:
image

File: examples/data-objects/canvas/test/canvas.test.ts [hunk 2]
```diff
@@ -55,17 +55,26 @@ describe("canvas", () => {
 				}
 				return context.getImageData(0, 0, width, height).data;
 			});
-			if (imgs[0].length == 0) {
+			const firstImg = imgs[0];
```
Comment (ChumpChief) [inline examples/data-objects/canvas/test/canvas.test.ts:58]: I just made a change here that will conflict - suggest just disabling the rule for the package for now.

**Review by noencke — APPROVED** [review]
Approved for packages/dds/tree/*

**Review by ChumpChief — CHANGES_REQUESTED** [review]
Please also revert the changes for collaborative-text and todo, then should be good.

**Review by ChumpChief — APPROVED** [review]
Thanks!

**Review by Josmithr — APPROVED** [review]

**Review by Abe27342 — APPROVED** [review]

**Pull request merged**